python/lib/core/dmod/core/events/router.py

```diff
@@ -0,0 +1,135 @@
+
+        for event_name, event_handlers in handlers.items():
+            self.register_handler(event_name, event_handlers)
+
+    def register_event(self, event: str, signature: SIGNATURE) -> EventRouter:
+        """
+        Register a handler signature with an event
+
+        Args:
+            event: The name of the event to register
+            signature: The expected signature for functions that will handle the event
+
+        Returns:
+            The updated EventRouter
+        """
+        if event in self.__events:
+            return self
+
+        if isinstance(signature, typing.Callable):
+            signature: Signature = Signature.from_function(signature)
+        elif isinstance(signature, inspect.Signature):
+            signature: Signature = Signature.from_signature(signature)
+        elif isinstance(signature, typing.Sequence):
+            signature = Signature(signature)
+
+        self.__events[event] = EventFunctionGroup(signature, allow_errors=self.__allow_errors)
+        return self
+
+    def register_handler(
+        self,
+        event: str,
+        handler: typing.Union[typing.Callable, typing.Sequence[typing.Callable]]
+    ) -> EventRouter:
+        if not handler:
+            return self
+
+        if not isinstance(handler, typing.Sequence):
+            handler = list(handler) if isinstance(handler, typing.Iterable) else [handler]
+
+        if not handler:
+            return self
+
+        if event not in self.__events:
+            logging.debug(
+                f"There is no registered event for '{event}' - "
+                f"a new event is being registered but the required signatures may be incorrect."
+            )
+            self.register_event(event, handler[0])
+
+        invalid_functions = list()
+
+        for event_handler in handler:
+            self.__events[event].add_function(event_handler, invalid_functions, allow_errors=self.__allow_errors)
+
+        if len(invalid_functions) > 0:
+            raise ValueError(
+                f"The following handlers cannot be added for the '{event}' event: {', '.join(invalid_functions)}"
+            )
+
+        return self
+
+    def __call__(self, event: typing.Union[str, Event], *args, **kwargs) -> typing.List[typing.Awaitable]:
+        if isinstance(event, str):
+            event = Event(event_name=event)
+
+        if not (self.__fail_on_missing_event or event.event_name in self.__events):
+            return list()
+        elif event.event_name not in self.__events:
+            raise ValueError(f"There are no registered handlers for the '{event.event_name}' event")
+
+        return self.__events[event.event_name](event, *args, **kwargs)
+
+    async def fire(self, event: typing.Union[str, Event], *args, **kwargs):
+        if isinstance(event, str):
+            event = Event(event_name=event)
+
+        if self.__fail_on_missing_event and event.event_name not in self.__events:
```

Comment:
Is this supposed to be?
```suggestion
        if not self.__fail_on_missing_event and event.event_name not in self.__events:
```

Reply:
Great catch. I made the change.